src/main/java/uk/ac/ic/wlgitbridge/bridge/context/ContextStore.java

```diff
@@ -0,0 +1,174 @@
+    Log.info("Initialize ContextStore");
+    if (ContextStore.instance != null) {
+        Log.error("ContextStore already initialized");
+      throw new ContextStoreAlreadyInitialisedException();
+    }
+    ContextStore.instance = new ContextStore(contextFactory);
+  }
+
+  /**
+   * Reset the ContextStore, only to be used in tests,
+   * hence the __scary__ underscores
+   */
+  public static void __Reset__() {
+    Log.info("Reset ContextStore");
+    if (ContextStore.instance != null) {
+      ContextStore.getInstance().stop();
+      ContextStore.instance = null;
+    }
+  }
+
+  public static ContextStore getInstance()  {
+    if (ContextStore.instance == null) {
+      throw new RuntimeException("ContextStore not initialized");
+    }
+    return ContextStore.instance;
+  }
+
+  /**
+   * This method takes a lambda, and runs it in a
+   * context with the project lock taken, and automatically
+   * cleans up at the end, returning a Pair<R, Exception>
+   *
+   * This pattern is used to get around a limitation of
+   * how lambdas interact with the exception system.
+   *
+   * @see uk.ac.ic.wlgitbridge.bridge.Bridge for usage.
+   *
+   * @param projectName : String project name
+   * @param f : A lambda, accepting a context as input,
+   *            and returning a Pair<R, Exception>
+   * @param <R> : The return type of the overall expression
+   * @return Pair<R, Exception>
+   */
+  public static <R> Pair<R, Exception> inContextWithLock(
+      String projectName,
+      Function<ProjectContext, Pair<R, Exception>> f
+  ) {
+    ContextStore instance = ContextStore.getInstance();
+    instance.ensureThreadLocalStorage();
+    if (instance.localContexts.get().get(projectName) != null) {
+      throw new RuntimeException("Context already exists");
+    }
+    ProjectContext context = instance.getContextForProject(projectName);
+    try (ProjectLock lock = context.getLock()) {
+      lock.lock();
+      Pair<R, Exception> result = f.apply(context);
+      lock.success();
+      return result;
+    } catch (Exception e) {
+      Log.error("[{}] Exception in context with lock", projectName, e);
+      return new Pair(null, e);
+    } finally {
+      context.close();
+      instance.localContexts.get().remove(projectName);
+    }
+  }
+
+  /**
+   * Cleanup contexts, and pass on the close signal.
+   * Essential for cleanly shutting down the server
+   */
+  public synchronized void stop() {
+    Log.info("Stopping ContextStore");
+    this.clearLocalContexts();
+    if (this.projectContextFactory != null) {
+      this.projectContextFactory.close();
+    }
+  }
+
+  private void ensureThreadLocalStorage() {
+    if (this.localContexts.get() == null) {
+      HashMap<String, ProjectContext> contextMap = new HashMap<String, ProjectContext>();
+      this.localContexts.set(contextMap);
+    }
+  }
+
+  private synchronized ProjectContext getFromThreadLocalStorage(String projectName) {
```

Comment:
Similar to `ensureThreadLocalStorage`, it doesn't look like this needs to be synchronized since it only accesses a thread-local variable.